http/server.lua

```diff
@@ -1295,11 +1296,80 @@ local function url_for_httpd(httpd, name, args, query)
     end
 end
 
+local function create_ssl_ctx(host, port, ssl_opts, use_tls)
+    if use_tls ~= true then
+        return nil
+    end
+
+    if not sslsocket_supported then
+        error("ssl socket is not supported")
+    end
+
+    local ok, ctx = pcall(sslsocket.ctx, sslsocket.tls_server_method())
+    if ok ~= true then
+        error(ctx)
+    end
+
+    local rc = sslsocket.ctx_use_private_key_file(ctx, ssl_opts.ssl_key_file,
+        ssl_opts.ssl_password, ssl_opts.ssl_password_file)
+    if rc == false then
+        errorf(
+            "Can't start server on %s:%s: %s %s",
+            host, port, 'Private key is invalid or password mismatch', ssl_opts.ssl_key_file
+        )
+    end
+
+    rc = sslsocket.ctx_use_certificate_file(ctx, ssl_opts.ssl_cert_file)
+    if rc == false then
+        errorf(
+            "Can't start server on %s:%s: %s %s",
+            host, port, 'Certificate is invalid', ssl_opts.ssl_cert_file
+        )
+    end
+
+    if ssl_opts.ssl_ca_file ~= nil then
+        rc = sslsocket.ctx_load_verify_locations(ctx, ssl_opts.ssl_ca_file)
+        if rc == false then
+            errorf(
+                "Can't start server on %s:%s: %s",
+                host, port, 'CA file is invalid'
+            )
+        end
+
+        sslsocket.ctx_set_verify(ctx, 0x01 + 0x02)
+    end
+
+    if ssl_opts.ssl_ciphers ~= nil then
+        rc = sslsocket.ctx_set_cipher_list(ctx, ssl_opts.ssl_ciphers)
+        if rc == false then
+            errorf(
+                "Can't start server on %s:%s: %s",
+                host, port, 'Ciphers is invalid'
+            )
+        end
+    end
+
+    return ctx, sslsocket.tcp_server
+end
+
 local function httpd_start(self)
     if type(self) ~= 'table' then
         error("httpd: usage: httpd:start()")
     end
 
+    local ssl_ctx = create_ssl_ctx(self.host, self.port, {
+        ssl_cert_file = self.options.ssl_cert_file,
+        ssl_key_file = self.options.ssl_key_file,
+        ssl_password = self.options.ssl_password,
+        ssl_password_file = self.options.ssl_password_file,
+        ssl_ca_file = self.options.ssl_ca_file,
+        ssl_ciphers = self.options.ssl_ciphers,
+    }, self.use_tls)
+
+    if ssl_ctx ~= nil then
+        self.tcp_server_f = sslsocket.tcp_server
```

Comment:
tcp_server_f could be used to replace an internal server by a user, see:
https://github.com/tarantool/http/blob/master/README.md#using-a-special-socket
https://github.com/tarantool/http/blob/22917310e6e6c1ff85bd6338b498984d7cfb67ec/http/server.lua#L1497-L1498

In this case we remove the ability.

I think the code should be moved to the `new` method, something like:

```Lua
if ssl.use_tls then
    self.tcp_server_f = function(host, port, timeout)
        local ssl_ctx = create_ssl_ctx(self.host, self.port, {
            ssl_cert_file = self.options.ssl_cert_file,
            ssl_key_file = self.options.ssl_key_file,
            ssl_password = self.options.ssl_password,
            ssl_password_file = self.options.ssl_password_file,
            ssl_ca_file = self.options.ssl_ca_file,
            ssl_ciphers = self.options.ssl_ciphers,
        })
        return sslsocket.tcp_server(host, port, timeout, ssl_ctx)
    end
else
    self.tcp_server_f = socket.tcp_server
end
```

Reply:
Refactored.